src/guest_memory.rs

```diff
@@ -0,0 +1,237 @@
+    /// Incomplete write
+    ShortWrite {
+        expected: GuestAddressValue,
+        completed: GuestAddressValue,
+    },
+    /// Incomplete read
+    ShortRead {
+        expected: GuestAddressValue,
+        completed: GuestAddressValue,
+    },
+
+    /// Syscall returned the given error.
+    SystemCallFailed(io::Error),
+    /// Requested backend address is out of range.
+    InvalidBackendAddress,
+    /// Requested offset is out of range.
+    InvalidBackendOffset,
+    /// Backend doesn't support the operation.
+    InvalidBackendOperation,
+}
+
+impl std::error::Error for Error {}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Guest memory error: ")?;
+        match self {
+            Error::NoMemoryRegion => write!(f, "no region found"),
+            Error::MemoryRegionOverlap => write!(f, "overlapping memory regions"),
+            Error::InvalidGuestAddress(addr) => {
+                write!(f, "invalid guest address {}", addr.raw_value())
+            }
+            Error::InvalidGuestAddressRange(base, size) => write!(
+                f,
+                "invalid address range, base {}/size {}",
+                base.raw_value(),
+                size,
+            ),
+            Error::WriteToMemory(_) => write!(f, "couldn't write to memory"),
+            Error::ReadFromMemory(_) => write!(f, "couldn't read from memory"),
+            Error::WriteToTarget(_) => write!(f, "couldn't write to target"),
+            Error::ReadFromSource(_) => write!(f, "couldn't read from source"),
+            Error::ShortWrite {
+                expected,
+                completed,
+            } => write!(
+                f,
+                "incomplete write of {} instead of {} bytes",
+                completed, expected,
+            ),
+            Error::ShortRead {
+                expected,
+                completed,
+            } => write!(
+                f,
+                "incomplete read of {} instead of {} bytes",
+                completed, expected,
+            ),
+            Error::SystemCallFailed(e) => write!(f, "syscall failed due to {}", e),
+            Error::InvalidBackendAddress => write!(f, "invalid backend address"),
+            Error::InvalidBackendOffset => write!(f, "invalid backend offset"),
+            Error::InvalidBackendOperation => write!(f, "invalid backend operation"),
+        }
+    }
+}
+
+/// Represents a guest physical address (GPA).
+///
+/// Notes:
+/// - On ARM64, a 32-bit hypervisor may be used to support a 64-bit guest. For simplicity,
+/// u64 is used to store the the raw value no matter the guest a 32-bit or 64-bit virtual machine.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GuestAddress(pub u64);
+impl_address_ops!(GuestAddress, u64);
+
+/// Type of the raw value stored in a GuestAddress object.
+pub type GuestAddressValue = <GuestAddress as AddressValue>::V;
```

Comment:
Seems that rust prefers associated types against generic parameters, because it needs more typing when using a trait with generic parameter and sometimes the code with generic parameters looks ugly.

Reply:
Ok, maybe we can pick up this discussion in another PR, since there are some nuances here, if ppl aren't terribly annoyed by it already :D